### ISSUE TITLE
Cannot pass test03

### ISSUE THREAD

**Issue by Yanksi** [Digital-Humans-23/a3#12]
Is it possible to provide more information on how the smoothness loss is actually compted for Ex3? I have spent a lot of time just for tuning the parameters to match my final result with the reference result with no sucess.

In test03, the velocity computed by our code is compared directly against the reference velocity, which could be requiring us to look for the exactly same implementation as the reference one. It would take a lot of time to explore every possible configuration on a cpu-only machine.

**Comment by yz-cnsdqz** [comment]
I tested the CPU-only version takes less than 1 second per iteration, so it should be fast without GPU. If your code is too slow, then please replace your nested loops into tensor operations.

Another trick is to replace the *sum* with *mean*, try *l1* or *l2* norms, etc. Here is a discussion that happened before: https://github.com/Digital-Humans-23/a3/issues/9

**Comment by Joshua31415** [comment]
Could you perhaps provide any additional hints as to how the threshold from the test can be reached? I've tried for a fairly long time by now with different hyperparameters, different formulations of the temporal smoothing losses (e.g. sum, mean, l1, l2, matrix difference for rotation, proper angle difference for rotations, ...) but am not able to pass the test. The lowest error I achieved was 0.09 doing so.

As the test is not actually testing smoothness of the velocity but how close our velocity is to the reference it is not really intuitive how we can pass the test since a smoother velocity is not necessarily closer to the reference.

**Comment by Yanksi** [comment]
> 

The training procedure takes about 0.5 sec per iteration. But for 2000 iterations, it is still about 20 minutes to train for a single configuration. Even I finally managed to get my code to pass to test03 by doing automatic search in the parameter space  by brute force, I honestly find this part doesn't make much sense to be part of the homework.

According to the final configuration I derived, using euler angle as rotation loss, which should be a much more mathematically sound way than simply using l1/l2 norm, was not part of the solution.

Therefore, please consider changing the evalution method of ex3 so that we can stop spending time on this simple parameter tuning game.

**Comment by yz-cnsdqz** [comment]
@Joshua31415  @Yanksi 

Without the smoothness term, a possible solution is the per-frame error is low but with visible jittering, which causes the estimated velocity is far away from the pseudo ground truth velocity. From the math perspective, the smoothness term makes the solution function from C0 to C1. 

Besides penalising the joint location velocity, you can also penalise the joint rotation velocity together. Have you tried penalising the velocity of the 6D continuous rotation representation or quaternion? Matrix rotation difference also should not only be $|A-B|$, but also can be $|A^T B-I|$. Euler angle is not the optimal solution, because it would suffer from the gimbal lock, and its conversion to the rotation matrix is not unique.

FYI, if 2000 iteration takes long, you can also think of reducing the iterations. Hyper-parameters are important in practice rather than a game. 0.08 is high enough, because its unit is 'meter/second'. For this a3, brute-force hyper-parameter search is unnecessary.

**Comment by Joshua31415** [comment]
Thanks a lot for the elaboration! Using the 6D representation I was able to pass the test without any further tuning of parameters.

**Comment by Pablo-Paniagua** [comment]
Is it acceptable to increase the number of iterations to further improve our final solution or should they be kept at 2k?

The test passes just fine with 2k, but given more time I see how it reduces a bit further (not such a significant increase but it does improve).

**Comment by yz-cnsdqz** [comment]
@Pablo-Paniagua 
yes it is acceptable. As mentioned in https://github.com/Digital-Humans-23/a3/issues/5, you are free to modify the code to pass a3.
